<a href="https://colab.research.google.com/github/Srushtirodge/Capstone_Project_ML-Yes_Bank_Stock_Closing_Price_Prediction-/blob/main/Capstone_Project_ML(Yes_Bank_Stock_Closing_Price_Prediction).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Yes_Bank_Stock_Closing_Price_Prediction(Capstone_Project_Supervised_ML_Regression)**

##### **Project Type**    - Regression
##### **Contribution**    - Srushti Rodge

# **`Project Summary -`**


Yes_Bank_Stock_Closing_Price_Prediction(Capstone_Project_Supervised_ML_Regression)
Project Type - Regression
Contribution - Team
Team Member 1 - Vibuti Gupta
Team Member 2 - Rudraashish Sengupta
Project Summary -
Yes_Bank_Logo-01.png

This project aims to address the challenge of predicting the closing price of Yes Bank's stock, which is a crucial task for stakeholders, investors, and market participants. Yes Bank, as a prominent private sector bank in India, has encountered significant hurdles in recent years, including a substantial number of bad loans and involvement in fraud cases. These challenges have led to regulatory intervention by the Reserve Bank of India, making the prediction of Yes Bank's stock prices complex and uncertain.

To tackle this challenge, the project utilizes a comprehensive dataset comprising monthly stock price data since the bank's inception. The dataset includes various important metrics such as the closing, starting, highest, and lowest prices for each month. The objective is to develop predictive models that can effectively capture the intricate dynamics and trends in Yes Bank's stock prices, taking into account the turbulent events and uncertainties associated with the bank's performance.

The project involves applying various modeling techniques, including time series models regression methods. The goal is to evaluate the performance of these models in accurately forecasting the closing price of Yes Bank's stock. Additionally, the models will be assessed on their ability to incorporate the impact of significant events, such as fraud cases involving the bank's founders or regulatory interventions by the Reserve Bank of India.

By successfully predicting the closing price of Yes Bank's stock, this project can provide valuable insights and assist stakeholders in making informed decisions related to their investments in the bank. It aims to navigate the complexities and uncertainties surrounding Yes Bank's stock prices, ultimately contributing to the understanding of its financial performance and aiding in effective decision-making.

# **`GitHub Link -`**

# **`Problem Statement`**

The main objective of this project is to develop a robust and accurate predictive model that can effectively forecast the closing price of Yes Bank's stock. The challenge lies in understanding and capturing the complex dynamics and trends in the stock prices, considering various factors such as the historical trend of an increasing price followed by a sudden decline after 2018.

One of the key challenges in developing the predictive model is addressing the issue of multicollinearity present in the dataset. Multicollinearity occurs when there is a high correlation between independent variables, which can lead to difficulties in interpreting the model and can affect the accuracy of the predictions. Therefore, the model should incorporate techniques to handle multicollinearity and ensure that the independent variables are appropriately considered in the prediction process.

Furthermore, the model should account for significant events that have had an impact on Yes Bank's stock performance. This includes events such as fraud cases involving the bank's founders and regulatory interventions by the Reserve Bank of India. These events can significantly influence the stock prices, and it is crucial for the predictive model to capture and reflect their effects accurately.

In terms of performance, the model should aim for a high level of accuracy in forecasting the closing price of Yes Bank's stock. The 99% accuracy achieved by the K-Nearest Neighbors (KNN) Regression model serves as a benchmark, indicating the target accuracy that the developed model should strive to achieve. By achieving high accuracy, the predictive model can provide valuable insights to stakeholders, investors, and market participants, enabling them to make informed decisions and effectively manage their investments in Yes Bank's stock.

Overall, this project seeks to develop a predictive model that addresses the complexities and challenges associated with forecasting Yes Bank's stock prices. The ultimate goal is to provide stakeholders with a reliable tool that can enhance their understanding of the stock's future performance and support them in making well-informed investment decisions.

## ***`1. Know Your Data`***


### `Import Libraries`

In [1]:
# Import necessary libraries

# Import NumPy for numerical computations
import numpy as np

# Import Pandas for data manipulation and analysis
import pandas as pd

# Import Matplotlib for basic data visualization
import matplotlib.pyplot as plt

# Import Seaborn for advanced statistical visualizations
import seaborn as sns

# Import Plotly graph objects for interactive visualizations
import plotly.graph_objects as go

# Import the datetime module for working with dates and times
from datetime import datetime

# Import warnings module to ignore potential warnings
import warnings

### `Dataset Loading`

In [2]:
# Load Dataset
# This code imports the `drive` library from Google Colab.
# The 'drive' library allows us to mount our Google Drive to our Colab notebook.
# The 'mount()' function mounts our Google Drive to the '/content/drive' directory.
from google.colab import drive
drive.mount('/content/drive')


Mounted at /content/drive


In [3]:
# Load the stock price data from the CSV file
# This code reads the CSV file 'data_YesBank_StockPrices.csv' from the '/content/drive/MyDrive/capstone project-Supervised ML-Regression/Data File' directory.
# The 'pd.read_csv()' function reads the CSV file and returns a Pandas DataFrame.
stock_df = pd.read_csv('/content/drive/MyDrive/project module 6/data_YesBank_StockPrices.csv')

### `Dataset First View`

In [4]:
# Display the first few rows of the DataFrame with a cool color gradient
# This code uses the 'head()' method to display the first few rows of the DataFrame.
# The 'style.background_gradient()' method applies a color gradient to the background of the DataFrame.
# The 'cmap='cool'' argument specifies the colormap to use.
stock_df.head().style.background_gradient(cmap='cool')

### `Dataset Rows & Columns count`

In [5]:
print("The shape of the DataFrame is:", stock_df.shape)

The shape of the DataFrame is: (185, 5)


### `Dataset Information`

In [6]:
print("The information about the DataFrame is:", stock_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185 entries, 0 to 184
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Date    185 non-null    object 
 1   Open    185 non-null    float64
 2   High    185 non-null    float64
 3   Low     185 non-null    float64
 4   Close   185 non-null    float64
dtypes: float64(4), object(1)
memory usage: 7.4+ KB
The information about the DataFrame is: None


From the above analysis, we can conclude that:




*   The shape of our dataset is 185 rows and 5 columns. This means that the dataset has 185 data points, and each data point has 5 features. The data points represent the stock prices of YesBank for a period of time. The features represent the date, open price, high price, low price, close price, and volume of the stock.
*   Datatype of Date is given as object which we need to change that to DateTime. The Date column contains the date of the stock price. The data type of this column is currently object, which means that the values in this column are strings. We need to change the data type of this column to DateTime so that we can perform date-related operations on it, such as calculating the day of the week, the month, or the year.
*   Rest all features have float value as data point. The other 4 columns in the dataset contain floating-point numbers. These numbers represent the open price, high price, low price, and close price of the stock. The volume column contains the number of shares traded on a given day.

#### `Duplicate Values`

In [8]:
number_of_duplicates = len(stock_df[stock_df.duplicated()])

# Print the number of duplicate rows
print("The number of duplicate rows is:", number_of_duplicates)

The number of duplicate rows is: 0


#### `Missing Values/Null Values`

In [9]:
missing_values_count = stock_df.isnull().sum()

# Print the number of missing values in each column
print("The number of missing values in each column is:","\n", missing_values_count)

The number of missing values in each column is: 
 Date     0
Open     0
High     0
Low      0
Close    0
dtype: int64


## ***`2. Understanding Your Variables`***

In [10]:
column_names = stock_df.columns

# Print the column names
print("The column names of the DataFrame are:", column_names)

The column names of the DataFrame are: Index(['Date', 'Open', 'High', 'Low', 'Close'], dtype='object')


In [11]:
summary_statistics = stock_df.describe(include='all')

# Print the summary statistics
print("The summary statistics for all columns in the DataFrame are:","\n","--------------------------------------------------------------","\n",summary_statistics)

The summary statistics for all columns in the DataFrame are: 
 -------------------------------------------------------------- 
           Date        Open        High         Low       Close
count      185  185.000000  185.000000  185.000000  185.000000
unique     185         NaN         NaN         NaN         NaN
top     Jul-05         NaN         NaN         NaN         NaN
freq         1         NaN         NaN         NaN         NaN
mean       NaN  105.541405  116.104324   94.947838  105.204703
std        NaN   98.879850  106.333497   91.219415   98.583153
min        NaN   10.000000   11.240000    5.550000    9.980000
25%        NaN   33.800000   36.140000   28.510000   33.450000
50%        NaN   62.980000   72.550000   58.000000   62.540000
75%        NaN  153.000000  169.190000  138.350000  153.300000
max        NaN  369.950000  404.000000  345.500000  367.900000


### `Variables Description`

The dataset consists of monthly observations of Yes Bank stock prices since its listing on the stock exchange. The dataset includes the following features:

Date: This indicates the specific month for which the stock price is recorded.
Open: This represents the price of the stock at the beginning of the trading day when the stock exchange opens.
High: This indicates the highest price reached by the stock during the given month.
Low: This indicates the lowest price reached by the stock during the given month
Close: This represents the price of the stock at the end of the trading day when the stock exchange closes.
The dataset provides a comprehensive overview of the monthly performance of Yes Bank stock, including the opening, highest, lowest, and closing prices for each month since its listing on the stock exchange.

##  ***`3. Data Wrangling`***

Since the dataset does not contain any duplicate or null values, there is no need to perform any operations to treat them. The dataset is already clean in this regard. We can now proceed with the next step, which involves detecting and dealing with outliers.

Outliers are data points that significantly deviate from the majority of the observations and may have a disproportionate impact on analysis results. Detecting outliers is important to ensure accurate and reliable analysis.

To detect outliers in the dataset, various statistical techniques can be employed, such as the use of z-scores, interquartile range (IQR), or box plots. These methods help identify data points that fall outside a certain threshold or are significantly different from the rest of the data.

Once outliers are identified, there are different approaches to deal with them. One common approach is to remove the outliers from the dataset, assuming they are erroneous or represent extreme values that may skew the analysis. Another approach is to replace the outliers with more representative values, such as the mean, median, or a value obtained through interpolation or extrapolation.

In addition to outlier detection, it is necessary to change the data type of the datetime column from object to datetime format. This conversion allows for more effective handling and manipulation of dates and times. It enables various time-based operations, such as sorting, filtering, and extracting specific time intervals, which are crucial for analyzing and visualizing the stock price data over time.

In [13]:
# Check Unique Values for each variable.

for variable in stock_df.columns:
  print(f"The unique values for the '{variable}' variable are:\n\n {stock_df[variable].unique()}\n\n")

The unique values for the 'Date' variable are:

 ['Jul-05' 'Aug-05' 'Sep-05' 'Oct-05' 'Nov-05' 'Dec-05' 'Jan-06' 'Feb-06'
 'Mar-06' 'Apr-06' 'May-06' 'Jun-06' 'Jul-06' 'Aug-06' 'Sep-06' 'Oct-06'
 'Nov-06' 'Dec-06' 'Jan-07' 'Feb-07' 'Mar-07' 'Apr-07' 'May-07' 'Jun-07'
 'Jul-07' 'Aug-07' 'Sep-07' 'Oct-07' 'Nov-07' 'Dec-07' 'Jan-08' 'Feb-08'
 'Mar-08' 'Apr-08' 'May-08' 'Jun-08' 'Jul-08' 'Aug-08' 'Sep-08' 'Oct-08'
 'Nov-08' 'Dec-08' 'Jan-09' 'Feb-09' 'Mar-09' 'Apr-09' 'May-09' 'Jun-09'
 'Jul-09' 'Aug-09' 'Sep-09' 'Oct-09' 'Nov-09' 'Dec-09' 'Jan-10' 'Feb-10'
 'Mar-10' 'Apr-10' 'May-10' 'Jun-10' 'Jul-10' 'Aug-10' 'Sep-10' 'Oct-10'
 'Nov-10' 'Dec-10' 'Jan-11' 'Feb-11' 'Mar-11' 'Apr-11' 'May-11' 'Jun-11'
 'Jul-11' 'Aug-11' 'Sep-11' 'Oct-11' 'Nov-11' 'Dec-11' 'Jan-12' 'Feb-12'
 'Mar-12' 'Apr-12' 'May-12' 'Jun-12' 'Jul-12' 'Aug-12' 'Sep-12' 'Oct-12'
 'Nov-12' 'Dec-12' 'Jan-13' 'Feb-13' 'Mar-13' 'Apr-13' 'May-13' 'Jun-13'
 'Jul-13' 'Aug-13' 'Sep-13' 'Oct-13' 'Nov-13' 'Dec-13' 'Jan-14' 'Feb-14'
 '

In [14]:
# Saving a copy of the original dataframe
og_df = stock_df.copy()

## **`Data Type Correction`**

Since the dataset does not contain any duplicate or null values, we do not need to perform any operations to treat them. We can proceed to outlier detection and dealing with them.

However, the datatype of the Date column is currently object. We need to change it to datetime format. This is because the Date column represents a date and time, and the object datatype is not sufficient to represent this type of data.

To change the datatype of the Date column, we can use the pd.to_datetime() function. For example, the following code would change the datatype of the Date column to datetime:

In [15]:
# Checking the exact datatype of the entries under the 'Date' column
type(stock_df['Date'][0])

str

In [23]:
# Changing date colunn datatype to datetime format.
from datetime import datetime

In [25]:
# parsing date which is string of format %b-%y to datetime (%b for Month as locale’s abbreviated name and %y for Year without century as a zero-padded decimal number.
stock_df['Date'] = stock_df['Date'].apply(lambda x: datetime.strftime(x, '%b-%y'))

In [26]:
print(stock_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 185 entries, 0 to 184
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Date    185 non-null    object 
 1   Open    185 non-null    float64
 2   High    185 non-null    float64
 3   Low     185 non-null    float64
 4   Close   185 non-null    float64
dtypes: float64(4), object(1)
memory usage: 7.4+ KB
None


In [27]:
#Check the datatype of the columns after changing datatype of date
stock_df.dtypes

Date      object
Open     float64
High     float64
Low      float64
Close    float64
dtype: object

In [28]:
# Setting the 'Date' column as the index
stock_df = stock_df.set_index('Date')

In [29]:
# Set the background color of the DataFrame to a gradient
# using 'style.background_gradient()' method
stock_df.head().style.background_gradient(cmap='cool')

In [30]:
dependent_variable = ['Close']
independent_variables = list(stock_df.columns[:-1])

Upon examining the provided dataframe, it becomes apparent that all the columns exclusively consist of numerical data. There is an absence of any categorical data in the dataset, which means that the information available for analysis primarily comprises quantitative values. This characteristic enables direct application of numerical calculations, statistical analyses, and modeling techniques to the dataset. The lack of categorical data simplifies data processing and ensures a streamlined approach when performing quantitative analyses.

Furthermore, during the examination of the dataset, it is evident that outliers are present. These outliers are data points that significantly deviate from the majority of the data. Before proceeding with modeling or conducting further analysis, it is crucial to address these outliers. Dealing with outliers involves assessing their impact on the data and making decisions regarding appropriate actions, such as removing or transforming them. By addressing the outliers, we can enhance the robustness and reliability of our models and analyses.

## ***`4. Data Vizualization, Storytelling & Experimenting with charts : Understand the relationships between variables`***

### Chart - 1 Line plot showcasing variations in each feature over the years

In [31]:
import plotly.express as px

# Plotting line graph wrt Date and Low prices
fig = px.line(og_df, x="Date", y="Low")

# Add additional traces for Open, Close, and High prices
fig.add_scatter(x=og_df['Date'], y=og_df['Open'], name="Open",
                line_color='lime', marker_color='hotpink', marker_size=10)
fig.add_scatter(x=og_df['Date'], y=og_df['Close'], name="Close",
                line_color='cyan', marker_color='magenta', marker_size=10)
fig.add_scatter(x=og_df['Date'], y=og_df['High'], name="High",
                line_color='gold', marker_color='deepskyblue', marker_size=10)
fig.add_scatter(x=og_df['Date'], y=og_df['Low'], name="Low",
                line_color='orange', marker_color='chartreuse', marker_size=10)

# Update the layout of the plot
fig.update_layout(
    title={'text': "Yes Bank Prices with Respect to Year", 'x': 0.5, 'y': 0.95, 'xanchor': 'center', 'yanchor': 'top', 'font': {'color': 'white'}},
    xaxis_title="Year",
    yaxis_title="Price",
    width=1400,
    height=800,
    plot_bgcolor='rgb(36, 40, 47)',  # Set the dark blue background color of the plot
    paper_bgcolor='rgb(51, 56, 66)',  # Set the dark blue background color of the paper area
    font_color='white',  # Set the font color to white
    legend=dict(x=0.02, y=0.98, bgcolor='rgba(255, 255, 255, 0.7)', bordercolor='gray', borderwidth=1, font={'color': 'white'}),
    margin=dict(l=50, r=50, t=100, b=50),
    xaxis=dict(tickangle=90)  # Rotate x-axis labels by 90 degrees
)
# Show the plot
fig.show()

#### `1. Why did you pick the specific chart?`

The specific chart chosen for this analysis is a combination of line and scatter plots. This chart type is suitable for visualizing the individual changes in Open, High, Low, and Close prices of the Yes Bank stock over time. By utilizing line plots, we can observe the overall trends and patterns, while scatter plots allow us to identify specific data points. The chart effectively presents the data by distinguishing each price variable with unique line colors and marker styles. The layout of the plot is optimized to include a centered title, clear axis labels, and an appropriate size. Additionally, the choice of color scheme and background enhances visual appeal and readability. Overall, this chart enables a comprehensive analysis of the Yes Bank stock prices, aiding in the identification of trends, patterns, and potential insights for informed decision-making.

#### `2. What is/are the insight(s) found from the chart?`

Indeed, the expected dip in the price variables after 2018 is prominently visible in the chart. The line graph shows a notable decrease in the Open, High, Low, and Close prices of the Yes Bank stock following the specified timeframe. This decline can be attributed to various factors, such as the impact of the Yes Bank fraud case involving Rana Kapoor, which adversely affected investor sentiment and confidence in the bank. The scatter plots further accentuate the dip, as they highlight individual data points that deviate significantly from the preceding upward trend. By visually representing the price variables over time, the chart effectively showcases the substantial decrease in prices after 2018, emphasizing the challenging period faced by Yes Bank and the subsequent decline in its stock value.

#### `3. Will the gained insights help creating a positive business impact?

The gained insights can potentially help create a positive business impact by providing valuable information for decision-making and strategic planning. By analyzing the Yes Bank stock price data and observing the significant dip after 2018, businesses and investors can adjust their strategies accordingly. These insights can guide them in making informed decisions about investing in Yes Bank or adjusting their existing holdings. Additionally, the insights can alert businesses to the need for diligent risk management practices and thorough due diligence when evaluating financial institutions.

Regarding insights leading to negative growth, the significant dip in the Yes Bank stock prices after 2018 can be seen as a negative growth trend. The decline in stock prices can be attributed to various factors, including the Yes Bank fraud case involving Rana Kapoor. This event had a detrimental impact on investor sentiment and eroded trust in the bank, resulting in a decrease in its stock value. The negative growth observed in this scenario highlights the importance of maintaining ethical practices, strong corporate governance, and transparency within financial institutions. It also underscores the potential consequences of fraud and misconduct on the overall growth and stability of a business.